File: src/main/java/com/hileco/cortex/analysis/processors/JumpUnreachableProcessor.java
package com.hileco.cortex.analysis.processors;

import com.hileco.cortex.analysis.Graph;
import com.hileco.cortex.analysis.edges.EdgeFlowMapping;
import com.hileco.cortex.instructions.debug.NOOP;

import java.util.HashSet;

public class JumpUnreachableProcessor implements Processor {

    private static final int PROGRAM_START = 0;

    @Override
    public void process(Graph graph) {
        EdgeFlowMapping.UTIL.findAny(graph).ifPresent(edgeFlowMapping -> {
            var targets = new HashSet<Integer>();
            targets.add(PROGRAM_START);
            targets.addAll(edgeFlowMapping.getFlowsToTarget().keySet());
            graph.getGraphBlocks().forEach(graphBlock -> graphBlock.getGraphNodes().stream().findFirst().ifPresent(startingNode -> {
                if (!targets.contains(startingNode.getLine())) {
                    graphBlock.getGraphNodes().forEach(graphNode -> graphNode.getInstruction().set(new NOOP()));
                }
            }));
        });
    }
}

File: src/main/java/com/hileco/cortex/analysis/processors/JumpUnreachableProcessor.kt
package com.hileco.cortex.analysis.processors

import com.hileco.cortex.analysis.Graph
import com.hileco.cortex.analysis.edges.EdgeFlowMapping
import com.hileco.cortex.instructions.debug.NOOP
import java.util.*

class JumpUnreachableProcessor : Processor {
    override fun process(graph: Graph) {
        graph.edgeMapping.get(EdgeFlowMapping::class.java).first().let {
            val targets = HashSet<Int>()
            targets.add(PROGRAM_START)
            targets.addAll(it.flowsToTarget.keys.filterNotNull())
            val unreachableBlocks = graph.graphBlocks.filterNot { graphBlock ->
                val startNode = graphBlock.graphNodes.first()
                targets.contains(startNode.line)
            }
            unreachableBlocks.forEach { graphBlock ->
                graphBlock.graphNodes.forEach { graphNode -> graphNode.instruction = NOOP() }
                graph.mergeUpwards(graphBlock)
            }
        }
    }

    companion object {
        private const val PROGRAM_START = 0
    }
}
